### PR TITLE
Feature/simulate sensor node

### PR DESCRIPTION
**Pull request by larsfroelich**
# Simulate sensor software
This PR allows running the sensor project in a "simulate" mode, skipping interactions with sensors, raspberry pi libraries and I/O pins.
Instead, random data is generated to simulate a functioning sensor node, mainly for testing other systems like server/database infrastucture,  or network/message communication.
It can also become useful as part of CI-tests (given that the whole sensor project is also dockerizable)

### Changes
Changed are limited to the `sensor/src` folder, nearly all files are affected there.

### Warning
Despite extensive reviewing, it is possible for one of the 300+ changed lines to introduce a new bug in the sensor project, so it is highly advisable to test-run code based on these changes in a suitable test environment first before broad deployment.

### CODE REVIEW

File: sensor/src/main.py [hunk 1]
```diff
@@ -27,14 +27,15 @@ def run() -> None:
     - Check for configuration update
     """
 
-    logger = utils.Logger(origin="main")
+    logger = utils.Logger(origin="main", print_to_console=True)
```
Comment (patrickjaigner) [inline sensor/src/main.py:30]: I guess this is only needed for the simulated device. I would add a if clause here.

File: sensor/src/procedures/calibration.py [hunk 2]
```diff
@@ -11,9 +11,11 @@ def __init__(
         self,
         config: custom_types.Config,
         hardware_interface: hardware.HardwareInterface,
+        simulate: bool = False,
     ) -> None:
         self.logger, self.config = utils.Logger(origin="calibration-procedure"), config
         self.hardware_interface = hardware_interface
+        self.simulate = simulate
```
Comment (patrickjaigner) [inline sensor/src/procedures/calibration.py:18]: Is this utilized for the simulation?

File: sensor/src/main.py (hunk 1 above)
Comment (larsfroelich) [reply sensor/src/main.py:30]: yes, good catch!

File: sensor/src/procedures/measurement.py
Comment (patrickjaigner) [inline sensor/src/procedures/measurement.py:1]: Is the simulation utilized here? I guess the hardware interface that are called in the procedure already know that they are simulated.

File: sensor/src/main.py (hunk 1 above)
Comment (larsfroelich) [reply sensor/src/main.py]: fixed it now, only logs to console if in simulate mode

File: sensor/src/procedures/calibration.py (hunk 2 above)
Comment (larsfroelich) [reply sensor/src/procedures/calibration.py:18]: no, it's essentially a stub that can be used in the future to "simulate" the calibration step, but for now calibration is skipped entirely during simulation, so we could also remove this from calibration.py entirely

File: sensor/src/procedures/measurement.py
Comment (larsfroelich) [reply sensor/src/procedures/measurement.py:1]: yes, it is also not used in here. it's a useful stub for simulating any measurement-specific code (in CI tests for example) but we could also remove it entirely to keep things cleaner

File: sensor/src/hardware/co2_sensor.py [hunk 3]
```diff
@@ -31,29 +32,32 @@ def __init__(
...
+        self.simulate = simulate
         self.logger.info("Starting initialization.")
         self.last_powerup_time: float = time.time()
 
-        # power pin to power up/down CO2 sensor
-        self.pin_factory = utils.get_gpio_pin_factory()
-        self.power_pin = gpiozero.OutputDevice(
-            pin=CO2_SENSOR_POWER_PIN_OUT, pin_factory=self.pin_factory
-        )
+        if not simulate:
```
Comment (patrickjaigner) [inline sensor/src/hardware/co2_sensor.py:47]: here you could probably:
if simulate:
    self.logger.info("Finished simulated initialization.")
    return

File: sensor/src/hardware/valves.py [hunk 4]
```diff
@@ -16,26 +16,29 @@ def __init__(
...
-        self.valves: dict[Literal[1, 2, 3, 4], gpiozero.OutputDevice] = {
-            1: gpiozero.OutputDevice(VALVE_PIN_1_OUT, pin_factory=self.pin_factory),
-            2: gpiozero.OutputDevice(VALVE_PIN_2_OUT, pin_factory=self.pin_factory),
-            3: gpiozero.OutputDevice(VALVE_PIN_3_OUT, pin_factory=self.pin_factory),
-            4: gpiozero.OutputDevice(VALVE_PIN_4_OUT, pin_factory=self.pin_factory),
-        }
-        self.active_input: Literal[1, 2, 3, 4] = self.config.measurement.valve_number
-        self.logger.info(f"Initialized with switching to valve: {self.active_input}")
-        self.set_active_input(self.active_input)
+        if not simulate:
```
Comment (patrickjaigner) [inline sensor/src/hardware/valves.py:30]: here also a if simulate: return is probably sufficient

**Review by patrickjaigner — COMMENTED** [review]
Looks good. We can probably make some smaller improvements before doing an integration session.

File: sensor/src/hardware/co2_sensor.py (hunk 3 above)
Comment (larsfroelich) [reply sensor/src/hardware/co2_sensor.py]: agreed, easier to read this way

File: sensor/src/hardware/valves.py (hunk 4 above)
Comment (larsfroelich) [reply sensor/src/hardware/valves.py]: have changed this to an early-return statement now

**Comment by larsfroelich** [comment]
I have updated all simulation-specific code to use an early-return statement rather than an `if not self.simulate:`
I would keep the `self.simulate = simulate`-stubs though, even in places where they aren't used/needed, just to keep things consistent and have the ability to add simulation-specific code there later.

**Review by patrickjaigner — APPROVED** [review]
Thank you for updating.

**Pull request merged**